### PR TITLE
ctrl+shift TreeView range selection

### PR DESCRIPTION
**Pull request by Meriipu**
issue #2479 

So I tried a bit and managed to get something working but it is an awful fix and I would hope there are better ways to do it, hopefully something native to pygtk that just requires changing an option. I am starting to doubt that native functionality exists, though this pull request as it stands now is a bit too awkward to be considered a solution in my opinion. 

On finding out whether it is a native gtk thing or not - firefox has the behaviour described in the issue with ctrl+shift for me in the bookmarks window (bookmarks->show all bookmarks) and qBittorrent has it in the torrent-list, although as far as I know qBittorrent is qt and not gtk.

Thunar does not seem to have it (it is gtk2 yes?), Firefox download list does not seem to have it, nor does my build of gimp in the open-files dialog.  

I looked at qltk/views.py, where the MultiDragTreeView seems to be the closest to defining something relevant (__block_selection), and otherwise inherits from native gtk things (treeview). After looking at the pygtk documentation I could not find any other way than manually setting the range. 

so I store the last selected treeview element, and when a ctrl+shift selection is made try to select the range from the stored element to the clicked element. 

awkward temporary flags to avoid endless recursion (self.__ctrl_shift in progress), as set_select_function apparently can not be changed (back to lambda *args: True) in the middle of a selection without segfaults? 

The actual selection logic might have some awkward elements too.

### CODE REVIEW

**Comment by Meriipu** [comment]
I tried to put together a test case as that might be useful even on its own but I can not get the helper _send_button_click_event by passing it coordinates of the rows of the widget as x,y and a shift modifier to actually select anything from what I can tell.

File: quodlibet/tests/helper.py [hunk 1]
```diff
@@ -143,6 +145,28 @@ def send_button_click(widget, button, primary=False, shift=False,
...
+        state |= Gdk.ModifierType.SHIFT_MASK
+
+    assert isinstance(widget, Gtk.Widget)
+    box = widget.get_cell_area(path, widget.get_column(0))
+    wx, wy = widget.convert_tree_to_widget_coords(box.x, box.y)
+    #width, height = widget.convert_tree_to_widget_coords(box.width, box.height)
+    wx += box.width // 2
+    wy += box.height // 2
+    handled = _send_button_click_event(widget, x=wx, y=wy, button=button, state=state)
+    # I did not get the parameter to enable stdout to work...
```
Comment (lazka) [inline quodlibet/tests/helper.py]: `py.test -s tests/test_qltk_views.py`

or `assert 0, "foo"` is also helpful

**Comment by lazka** [comment]
I don't know why it doesn't work. One would need to step through the gtk C code. My guess would be that we are missing many more event related events here. At least focus events, enter and leave events etc.

Maybe try to emulate the gtk behavior roughly by doing something like

```
    handled = _send_button_click_event(widget, x=wx, y=wy, button=button, state=state)
    if not handled:
        widget.get_selection().select_path(path)
```

**Comment by Meriipu** [comment]
Though if the behaviour that was to be tested was selecting elements with modifiers active (and then check whether the intended range was selected), emulating it that way would not really test it?

**Comment by lazka** [comment]
Yeah, you'd have to emulate that as well I guess. Event related code in gtk is sadly hard to test :(

**Comment by Meriipu** [comment]
this was pretty much a disaster so closing for now

**Pull request closed without merging**